pkg/chaosctl/common/exec.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2019 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"strconv"
+	"strings"
+
+	"google.golang.org/grpc/grpclog"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/remotecommand"
+	kubectlscheme "k8s.io/kubectl/pkg/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+// Exec executes certain command and returns the result
+// runtime-controller only support CRUD， use client-go client
+func Exec(ctx context.Context, pod v1.Pod, daemon v1.Pod, cmd string, c *kubernetes.Clientset) (string, error) {
+	out, err := exec(ctx, pod, daemon, cmd, c)
+
+	if err != nil {
+		// use daemon to enter namespace and execute command if command not found (which stream would failed)
+		if strings.Contains(err.Error(), "streaming remotecommand") {
+			outNs, errNs := nsEnterExec(ctx, err.Error(), pod, daemon, cmd, c)
+			if errNs == nil {
+				return outNs, nil
+			}
+			err = fmt.Errorf("%s\nnsenter also failed with: %s", err.Error(), errNs.Error())
+		}
+		return "", err
+	}
+
+	return out, nil
+}
+
+func exec(ctx context.Context, pod v1.Pod, daemon v1.Pod, cmd string, c *kubernetes.Clientset) (string, error) {
```

Comment:
Is the `daemon` a redundant argument here?

Reply:
My bad. Thx!